File: createOrder.js
require("dotenv").config();
const { PrivyClient } = require("@privy-io/node");
const { ethers } = require("ethers");
const {
  createAuthenticatedRequest,
  hexToPrivateKey,
} = require("./orderly-auth");

// Use native fetch in Node.js 18+ or fallback to node-fetch
let fetch;
if (typeof globalThis !== "undefined" && globalThis.fetch) {
  fetch = globalThis.fetch;
} else {
  try {
    fetch = require("node-fetch");
  } catch (e) {
    throw new Error(
      "fetch is not available. Please install node-fetch: npm install node-fetch"
    );
  }
}

// Configuration
const ORDERLY_API_URL = "https://api.orderly.org"; // or https://testnet-api.orderly.org for testnet
const BROKER_ID = "woofi_pro";

/**
 * Generate Orderly account ID
 * Based on Orderly documentation: accountId = keccak256(abi.encode(address, keccak256(abi.encodePacked(brokerId))))
 *
 * @param {string} address - User wallet address
 * @param {string} brokerId - Broker ID
 * @returns {string} - Account ID as bytes32 hex string
 */
function getAccountId(address, brokerId) {
  const abiCoder = ethers.AbiCoder.defaultAbiCoder();
  // keccak256(abi.encodePacked(brokerId)) - using solidityPackedKeccak256 for ABI-encoded string
  const brokerIdHash = ethers.solidityPackedKeccak256(["string"], [brokerId]);
  // keccak256(abi.encode(address, brokerIdHash))
  return ethers.keccak256(
    abiCoder.encode(["address", "bytes32"], [address, brokerIdHash])
  );
}

/**
 * Create an order on Orderly Network using Privy agentic wallet
 * Based on Orderly documentation: https://orderly.network/docs/build-on-omnichain/evm-api/restful-api/private/create-order
 *
 * @param {Object} options - Configuration options
 * @param {string} options.walletId - The Privy wallet ID (required)
 * @param {string} options.walletAddress - The wallet address (optional, will be fetched if not provided)
 * @param {string} options.symbol - Trading symbol (e.g., "PERP_ETH_USDC")
 * @param {string} options.orderType - Order type: "LIMIT", "MARKET", "IOC", "FOK", "POST_ONLY", "ASK", "BID"
 * @param {string} options.side - Order side: "BUY" or "SELL"
 * @param {number} [options.orderPrice] - Order price (required for LIMIT, IOC, FOK, POST_ONLY)
 * @param {number} [options.orderQuantity] - Order quantity in base currency
 * @param {number} [options.orderAmount] - Order amount in quote currency (for MARKET/BID/ASK orders)
 * @param {number} [options.visibleQuantity] - Visible quantity on orderbook (default: order_quantity)
 * @param {boolean} [options.reduceOnly] - Reduce only flag (default: false)
 * @param {number} [options.slippage] - Slippage tolerance for MARKET orders
 * @param {string} [options.clientOrderId] - Custom client order ID (36 chars max, unique)
 * @param {string} [options.orderTag] - Order tag
 * @param {number} [options.level] - Level for BID/ASK orders (0-4)
 * @param {boolean} [options.postOnlyAdjust] - Price adjustment for POST_ONLY orders
 * @returns {Promise<Object>} - Order creation result
 */
async function createOrder(options = {}) {
  try {
    // Validate environment variables
    const appId = process.env.PRIVY_APP_ID;
    const appSecret = process.env.PRIVY_APP_SECRET;

    if (!appId || !appSecret) {
      throw new Error(
        "Missing required environment variables. Please ensure the following are set in your .env file:\n" +
          "- PRIVY_APP_ID\n" +
          "- PRIVY_APP_SECRET"
      );
    }

    if (!options.walletId) {
      throw new Error("Wallet ID is required. Use --wallet-id <wallet_id>");
    }

    // Get Orderly credentials
    const orderlyKey = process.env.ORDERLY_KEY;
    if (!orderlyKey) {
      throw new Error(
        "ORDERLY_KEY environment variable is required (should be in format 'ed25519:...')"
      );
    }

    const orderlyPrivateKeyHex = process.env.ORDERLY_PRIVATE_KEY;
    if (!orderlyPrivateKeyHex) {
      throw new Error(
        "ORDERLY_PRIVATE_KEY environment variable is required (ed25519 private key in hex format)"
      );
    }

    // Convert hex string to Uint8Array
    const orderlyPrivateKey = hexToPrivateKey(orderlyPrivateKeyHex);

    // Initialize Privy client
    const privy = new PrivyClient({
      appId: appId,
      appSecret: appSecret,
    });

    // Configuration
    const brokerId = BROKER_ID;

    // Get wallet address if not provided
    let walletAddress = options.walletAddress;
    if (!walletAddress) {
      console.log("Fetching wallet details...");
      const wallet = await privy.wallets().get(options.walletId);
      walletAddress = wallet.address || wallet.addresses?.[0];
      if (!walletAddress) {
        throw new Error(
          "Could not determine wallet address from wallet object"
        );
      }
      console.log(`   Wallet Address: ${walletAddress}`);
    }

    // Derive Orderly account ID from wallet address and broker ID
    const accountId = getAccountId(walletAddress, brokerId);

    // Extract order parameters
    const {
      symbol,
      orderType,
      side,
      orderPrice,
      orderQuantity,
      orderAmount,
      visibleQuantity,
      reduceOnly = false,
      slippage,
      clientOrderId,
      orderTag,
      level,
      postOnlyAdjust,
    } = options;

    console.log("\nPreparing order creation...");
    console.log(`   Wallet ID: ${options.walletId}`);
    console.log(`   Wallet Address: ${walletAddress}`);
    console.log(`   Account ID: ${accountId}`);
    console.log(`   Symbol: ${symbol}`);
    console.log(`   Order Type: ${orderType}`);
    console.log(`   Side: ${side}`);

    // Validate required fields
    if (!symbol || !orderType || !side) {
      throw new Error(
        "Missing required fields: symbol, orderType, and side are required"
      );
    }

    // Validate order type
    const validOrderTypes = [
      "LIMIT",
      "MARKET",
      "IOC",
      "FOK",
      "POST_ONLY",
      "ASK",
      "BID",
    ];

    if (!validOrderTypes.includes(orderType.toUpperCase())) {
      throw new Error(
        `Invalid order type: ${orderType}. Must be one of: ${validOrderTypes.join(
          ", "
        )}`
      );
    }

    // Validate side
    if (!["BUY", "SELL"].includes(side.toUpperCase())) {
      throw new Error(`Invalid side: ${side}. Must be BUY or SELL`);
    }

    // Validate order_price requirement for certain order types
    const orderTypeUpper = orderType.toUpperCase();
    if (
      ["LIMIT", "IOC", "FOK", "POST_ONLY"].includes(orderTypeUpper) &&
      orderPrice === undefined
    ) {
      throw new Error(`orderPrice is required for ${orderType} orders`);
    }

    // Validate order_quantity or order_amount
    if (orderQuantity === undefined && orderAmount === undefined) {
      throw new Error("Either orderQuantity or orderAmount must be provided");
    }

    // Validate MARKET/BID/ASK order requirements
    if (["MARKET", "BID", "ASK"].includes(orderTypeUpper)) {
      if (side.toUpperCase() === "SELL" && orderAmount !== undefined) {
        throw new Error(
          "orderAmount is not supported for SELL orders with MARKET/BID/ASK order types"
        );
      }
      if (side.toUpperCase() === "BUY" && orderQuantity !== undefined) {
        throw new Error(
          "orderQuantity is not supported for BUY orders with MARKET/BID/ASK order types"
        );
      }
    }

    // Build request body
    const requestBody = {
      symbol: symbol,
      order_type: orderType.toUpperCase(),
      side: side.toUpperCase(),
    };

    // Add optional fields
    if (orderPrice !== undefined) {
      requestBody.order_price = orderPrice;
    }
    if (orderQuantity !== undefined) {
      requestBody.order_quantity = orderQuantity;
    }
    if (orderAmount !== undefined) {
      requestBody.order_amount = orderAmount;
    }
    if (visibleQuantity !== undefined) {
      requestBody.visible_quantity = visibleQuantity;
    }
    if (reduceOnly !== undefined) {
      requestBody.reduce_only = reduceOnly;
    }
    if (slippage !== undefined) {
      requestBody.slippage = slippage;
    }
    if (clientOrderId !== undefined) {
      requestBody.client_order_id = clientOrderId;
    }
    if (orderTag !== undefined) {
      requestBody.order_tag = orderTag;
    }
    if (level !== undefined) {
      requestBody.level = level;
    }
    if (postOnlyAdjust !== undefined) {
      requestBody.post_only_adjust = postOnlyAdjust;
    }

    console.log("\nOrder parameters:", JSON.stringify(requestBody, null, 2));

    const path = "/v1/order";
    const requestConfig = await createAuthenticatedRequest(
      "POST",
      path,
      requestBody,
      accountId,
      orderlyKey,
      orderlyPrivateKey
    );

    const response = await fetch(`${ORDERLY_API_URL}${path}`, requestConfig);
    const data = await response.json();

    // Check both HTTP status and success field in response body
    if (!response.ok) {
      throw new Error(`Failed to create order: ${JSON.stringify(data)}`);
    }

    if (!data.success) {
      throw new Error(`Order creation failed: ${JSON.stringify(data)}`);
    }

    console.log("\n✅ Order created successfully!");
    console.log("Response:", JSON.stringify(data, null, 2));

    return {
      success: true,
      data: data.data,
      orderId: data.data?.order_id,
      clientOrderId: data.data?.client_order_id,
      walletAddress: walletAddress,
      accountId: accountId,
    };
  } catch (error) {
    console.error("\n❌ Failed to create order");
    if (error.message) {
      console.error("Error:", error.message);
    } else {
      console.error(error);
    }
    throw error;
  }
}

/**
 * Main execution function
 */
async function main() {
  try {
    // Parse command line arguments
    const args = process.argv.slice(2);
    const options = {};

    // Simple argument parsing
    for (let i = 0; i < args.length; i++) {
      if (args[i] === "--wallet-id" && args[i + 1]) {
        options.walletId = args[i + 1];
        i++;
      } else if (args[i] === "--wallet-address" && args[i + 1]) {
        options.walletAddress = args[i + 1];
        i++;
      } else if (args[i] === "--symbol" && args[i + 1]) {
        options.symbol = args[i + 1];
        i++;
      } else if (args[i] === "--order-type" && args[i + 1]) {
        options.orderType = args[i + 1];
        i++;
      } else if (args[i] === "--side" && args[i + 1]) {
        options.side = args[i + 1];
        i++;
      } else if (args[i] === "--order-price" && args[i + 1]) {
        options.orderPrice = parseFloat(args[i + 1]);
        i++;
      } else if (args[i] === "--order-quantity" && args[i + 1]) {
        options.orderQuantity = parseFloat(args[i + 1]);
        i++;
      } else if (args[i] === "--order-amount" && args[i + 1]) {
        options.orderAmount = parseFloat(args[i + 1]);
        i++;
      } else if (args[i] === "--visible-quantity" && args[i + 1]) {
        options.visibleQuantity = parseFloat(args[i + 1]);
        i++;
      } else if (args[i] === "--reduce-only") {
        options.reduceOnly = true;
      } else if (args[i] === "--slippage" && args[i + 1]) {
        options.slippage = parseFloat(args[i + 1]);
        i++;
      } else if (args[i] === "--client-order-id" && args[i + 1]) {
        options.clientOrderId = args[i + 1];
        i++;
      } else if (args[i] === "--order-tag" && args[i + 1]) {
        options.orderTag = args[i + 1];
        i++;
      } else if (args[i] === "--level" && args[i + 1]) {
        options.level = parseInt(args[i + 1]);
        i++;
      } else if (args[i] === "--post-only-adjust") {
        options.postOnlyAdjust = true;
      } else if (args[i] === "--help" || args[i] === "-h") {
        console.log(`
Usage: node createOrder.js [options]

Required Options:
  --wallet-id <id>        Privy wallet ID to use (required)
  --symbol <symbol>       Trading symbol (e.g., "PERP_ETH_USDC") (required)
  --order-type <type>     Order type: LIMIT, MARKET, IOC, FOK, POST_ONLY, ASK, BID (required)
  --side <side>           Order side: BUY or SELL (required)

Optional Options:
  --wallet-address <addr>  Wallet address (optional, will be fetched if not provided)
  --order-price <price>    Order price (required for LIMIT/IOC/FOK/POST_ONLY)
  --order-quantity <qty>   Order quantity in base currency
  --order-amount <amount>  Order amount in quote currency (for MARKET/BID/ASK BUY orders)
  --visible-quantity <qty> Visible quantity on orderbook (default: order_quantity)
  --reduce-only            Reduce only flag (default: false)
  --slippage <slippage>    Slippage tolerance for MARKET orders
  --client-order-id <id>   Custom client order ID (36 chars max, unique)
  --order-tag <tag>        Order tag
  --level <level>          Level for BID/ASK orders (0-4)
  --post-only-adjust       Price adjustment for POST_ONLY orders
  --help, -h               Show this help message

Examples:
  # LIMIT BUY order
  node createOrder.js --wallet-id wal_xxx --symbol "PERP_ETH_USDC" --order-type LIMIT --side BUY --order-price 2000 --order-quantity 0.1

  # MARKET BUY order (using orderAmount)
  node createOrder.js --wallet-id wal_xxx --symbol "PERP_ETH_USDC" --order-type MARKET --side BUY --order-amount 100

  # LIMIT SELL order
  node createOrder.js --wallet-id wal_xxx --symbol "PERP_ETH_USDC" --order-type LIMIT --side SELL --order-price 2100 --order-quantity 0.05

  # LIMIT BUY order with reduce-only
  node createOrder.js --wallet-id wal_xxx --symbol "PERP_ETH_USDC" --order-type LIMIT --side BUY --order-price 2000 --order-quantity 0.1 --reduce-only

Environment Variables:
  PRIVY_APP_ID              Privy App ID (required)
  PRIVY_APP_SECRET          Privy App Secret (required)
  ORDERLY_KEY               Orderly public key (ed25519:...) (required)
  ORDERLY_PRIVATE_KEY       Orderly private key in hex format (required)

Note: Orderly account ID is automatically derived from wallet address and broker ID.
        `);
        process.exit(0);
      }
    }

    const result = await createOrder(options);

    console.log("\n📝 Order Summary:");
    console.log(`   Wallet Address: ${result.walletAddress}`);
    console.log(`   Account ID: ${result.accountId}`);
    console.log(`   Order ID: ${result.orderId}`);
    if (result.clientOrderId) {
      console.log(`   Client Order ID: ${result.clientOrderId}`);
    }

    process.exit(0);
  } catch (error) {
    console.error("\n❌ Failed to create order:", error.message);
    process.exit(1);
  }
}

// Run if executed directly
if (require.main === module) {
  main();
}

module.exports = { createOrder, getAccountId };

File: py/create_order.py
"""
Create an order on Orderly Network using Privy agentic wallet
Based on Orderly documentation: https://orderly.network/docs/build-on-omnichain/evm-api/restful-api/private/create-order
"""
import os
import sys
import argparse
from dotenv import load_dotenv
from web3 import Web3
from orderly_auth import create_authenticated_request, hex_to_private_key
import requests
from privy_utils import get_account_id, get_wallet_address
from orderly_constants import ORDERLY_API_URL, BROKER_ID

load_dotenv()


def create_order(
    wallet_id: str,
    symbol: str = None,
    order_type: str = None,
    side: str = None,
    order_price: float = None,
    order_quantity: float = None,
    order_amount: float = None,
    visible_quantity: float = None,
    reduce_only: bool = False,
    slippage: float = None,
    client_order_id: str = None,
    order_tag: str = None,
    level: int = None,
) -> dict:
    """Create an order on Orderly Network"""
    app_id = os.getenv("PRIVY_APP_ID")
    app_secret = os.getenv("PRIVY_APP_SECRET")
    orderly_key = os.getenv("ORDERLY_KEY")
    orderly_private_key_hex = os.getenv("ORDERLY_PRIVATE_KEY")
    
    if not app_id or not app_secret:
        raise ValueError("Missing PRIVY_APP_ID or PRIVY_APP_SECRET")
    if not wallet_id:
        raise ValueError("Wallet ID is required")
    if not orderly_key:
        raise ValueError("ORDERLY_KEY environment variable is required")
    if not orderly_private_key_hex:
        raise ValueError("ORDERLY_PRIVATE_KEY environment variable is required")
    
    orderly_private_key = hex_to_private_key(orderly_private_key_hex)
    
    print("Fetching wallet details...")
    wallet_address = get_wallet_address(wallet_id, app_id, app_secret)
    print(f"   Wallet Address: {wallet_address}")
    
    account_id = get_account_id(wallet_address, BROKER_ID)
    
    # Validate required fields
    if not symbol or not order_type or not side:
        raise ValueError("Missing required fields: symbol, orderType, and side are required")
    
    # Validate order type
    valid_order_types = ["LIMIT", "MARKET", "IOC", "FOK", "POST_ONLY", "ASK", "BID"]
    if order_type.upper() not in valid_order_types:
        raise ValueError(f"Invalid order type: {order_type}. Must be one of: {', '.join(valid_order_types)}")
    
    # Validate side
    if side.upper() not in ["BUY", "SELL"]:
        raise ValueError(f"Invalid side: {side}. Must be BUY or SELL")
    
    # Validate order_price requirement
    order_type_upper = order_type.upper()
    if order_type_upper in ["LIMIT", "IOC", "FOK", "POST_ONLY"] and order_price is None:
        raise ValueError(f"orderPrice is required for {order_type} orders")
    
    # Validate order_quantity or order_amount
    if order_quantity is None and order_amount is None:
        raise ValueError("Either orderQuantity or orderAmount must be provided")
    
    # Validate MARKET/BID/ASK order requirements
    if order_type_upper in ["MARKET", "BID", "ASK"]:
        if side.upper() == "SELL" and order_amount is not None:
            raise ValueError("orderAmount is not supported for SELL orders with MARKET/BID/ASK order types")
        if side.upper() == "BUY" and order_quantity is not None:
            raise ValueError("orderQuantity is not supported for BUY orders with MARKET/BID/ASK order types")
    
    print("\nPreparing order creation...")
    print(f"   Wallet ID: {wallet_id}")
    print(f"   Wallet Address: {wallet_address}")
    print(f"   Account ID: {account_id}")
    print(f"   Symbol: {symbol}")
    print(f"   Order Type: {order_type}")
    print(f"   Side: {side}")
    
    # Build request body
    request_body = {
        "symbol": symbol,
        "order_type": order_type.upper(),
        "side": side.upper(),
    }
    
    if order_price is not None:
        request_body["order_price"] = order_price
    if order_quantity is not None:
        request_body["order_quantity"] = order_quantity
    if order_amount is not None:
        request_body["order_amount"] = order_amount
    if visible_quantity is not None:
        request_body["visible_quantity"] = visible_quantity
    if reduce_only is not None:
        request_body["reduce_only"] = reduce_only
    if slippage is not None:
        request_body["slippage"] = slippage
    if client_order_id is not None:
        request_body["client_order_id"] = client_order_id
    if order_tag is not None:
        request_body["order_tag"] = order_tag
    if level is not None:
        request_body["level"] = level
    
    print(f"\nOrder parameters: {request_body}")
    
    path = "/v1/order"
    request_config = create_authenticated_request(
        "POST",
        path,
        request_body,
        account_id,
        orderly_key,
        orderly_private_key
    )
    
    response = requests.post(
        f"{ORDERLY_API_URL}{path}",
        headers=request_config["headers"],
        data=request_config["body"]
    )
    
    data = response.json()
    
    if not response.ok:
        raise Exception(f"Failed to create order: {data}")
    
    if not data.get("success"):
        raise Exception(f"Order creation failed: {data}")
    
    print("\n✅ Order created successfully!")
    print(f"Response: {data}")
    
    return {
        "success": True,
        "data": data.get("data"),
        "orderId": data.get("data", {}).get("order_id"),
        "clientOrderId": data.get("data", {}).get("client_order_id"),
        "walletAddress": wallet_address,
        "accountId": account_id,
    }


def main():
    parser = argparse.ArgumentParser(description="Create an order on Orderly Network")
    parser.add_argument("--wallet-id", required=True, help="Privy wallet ID to use (required)")
    parser.add_argument("--symbol", required=True, help="Trading symbol (e.g., 'PERP_ETH_USDC') (required)")
    parser.add_argument("--order-type", required=True, help="Order type: LIMIT, MARKET, IOC, FOK, POST_ONLY, ASK, BID (required)")
    parser.add_argument("--side", required=True, choices=["BUY", "SELL"], help="Order side: BUY or SELL (required)")
    parser.add_argument("--order-price", type=float, help="Order price (required for LIMIT/IOC/FOK/POST_ONLY)")
    parser.add_argument("--order-quantity", type=float, help="Order quantity in base currency")
    parser.add_argument("--order-amount", type=float, help="Order amount in quote currency (for MARKET/BID/ASK BUY orders)")
    parser.add_argument("--visible-quantity", type=float, help="Visible quantity on orderbook")
    parser.add_argument("--reduce-only", action="store_true", help="Reduce only flag")
    parser.add_argument("--slippage", type=float, help="Slippage tolerance for MARKET orders")
    parser.add_argument("--client-order-id", help="Custom client order ID (36 chars max, unique)")
    parser.add_argument("--order-tag", help="Order tag")
    parser.add_argument("--level", type=int, help="Level for BID/ASK orders (0-4)")
    
    args = parser.parse_args()
    
    try:
        result = create_order(
            wallet_id=args.wallet_id,
            symbol=args.symbol,
            order_type=args.order_type,
            side=args.side,
            order_price=args.order_price,
            order_quantity=args.order_quantity,
            order_amount=args.order_amount,
            visible_quantity=args.visible_quantity,
            reduce_only=args.reduce_only,
            slippage=args.slippage,
            client_order_id=args.client_order_id,
            order_tag=args.order_tag,
            level=args.level,
        )
        
        print("\n📝 Order Summary:")
        print(f"   Wallet Address: {result['walletAddress']}")
        print(f"   Account ID: {result['accountId']}")
        print(f"   Order ID: {result['orderId']}")
        if result.get("clientOrderId"):
            print(f"   Client Order ID: {result['clientOrderId']}")
        
        sys.exit(0)
    except Exception as error:
        print(f"\n❌ Failed to create order: {error}")
        sys.exit(1)


if __name__ == "__main__":
    main()

